### ISSUE TITLE
Support InheritsFrom association (Joined Table Inheritance)

### ISSUE THREAD

**Issue by ludydoo** [sequelize/sequelize#1243]
I am using sequelize but ran into the problem of it not supporting model inheritance.

I would suggest implementing an InheritsFrom association, that would be basically the same as the BelongsTo association. Though, when persisting a new object, sequelize  would chain-create all subclass objects and persist them first. It would also allow an object to share the getters/setters of its superclass instances.

```
  module.exports = (function() {
  var InheritsFrom = function(source, target, options) {
    this.associationType      = 'InheritsFrom'
    this.source               = source
    this.target               = target
    this.options              = options
    this.isSingleAssociation  = true
    this.isSelfAssociation    = false
    this.as                   = this.options.as
```

I figured I could add the InheritsFrom and Superclasses options to the dao-factory.js,
Also, I would add InheritsFromInstance and SuperClassInstance in the dao.js

```
  var DAOFactory = function(name, attributes, options) {
    this.options = Utils._.extend({
      timestamps: true,
      ...
      inheritsFrom : null
      superClasses : {}
    }, options || {})
```

Calling the dao get from a inherited dao would look for the key, and if it was not found, look for the key in the sub dao, until it finds the correct  getter. Also, if a getter/setter was called from a superclass dao, it would find the last inherited dao, and start to find the key from there. 

```
DAO.prototype.get = function (key) {
    if (key) {
      if (this._customGetters[key]) {
        return this._customGetters[key].call(this, key)
      }
      return this.dataValues[key]
    }

    if (this._hasCustomGetters) {
      var values = {}
        , key

      for (key in this._customGetters) {
        if (this._customGetters.hasOwnProperty(key)) {
          values[key] = this.get(key)
        }
      }

      for (key in this.dataValues) {
        if (!values.hasOwnProperty(key) && this.dataValues.hasOwnProperty(key)) {
          values[key] = this.dataValues[key]
        }
      }
      return values
    }

    //Here, check if the superclass has the get methods

    return this.dataValues
  }
```

I think I have to add some _customGetters and _customSetters in the dao-factory.js, but could somebody help me out and point me where in the codebase should I add such functionality? There is very few comments in there!

Please comment !

**Comment by neekey** [comment]
seems Node-ORM has implemented something similar [extendsTo](https://github.com/dresende/node-orm2/wiki/extendsTo)

**Comment by neekey** [comment]
Also I posted a question about [How do I implement a MySQL inheritance relations with a Node.js database ORM (like sequelizejs)?](http://www.quora.com/How-do-I-implement-a-MySQL-inheritance-relations-with-a-Node-js-database-ORM-like-sequelizejs?__snids__=970865882%2C970865881%2C970852055%2C970847456&__nsrc__=2), any good suggestion?

**Comment by mickhansen** [comment]
@neekey i've though about implementing some simple code-side inheritance (i.e. not postgres style table inheritance), it shouldn't be too tricky actually :)

**Comment by skvsree** [comment]
@mickhansen @ludydoo  is this already in core sequelize or is it available as a plugin ?

**Comment by mickhansen** [comment]
@skvsree Not available in core, per my last comment.

**Comment by neekey** [comment]
@mickhansen looking forward to this feature :+1:

**Comment by futurechan** [comment]
:+1: this would be really useful for me too!

**Comment by wadi-chemkhi** [comment]
Is there any hope for object inheritance mapping become a feature soon ?

**Comment by diosney** [comment]
+1 Hope to see landing soon.

Will allow to define common attributes for models.

**Comment by csrinaldi** [comment]
Some progress on this issue?

**Comment by janmeier** [comment]
@csrinaldi No progress so far, other than the discussions here - we still need a more formal spec and a descision about which of the proposed solutions to implement.

You are welcome to weigh in with your comments about which method you prefer

**Comment by VividVisions** [comment]
:+1: Would be great!

**Comment by naviapis** [comment]
+1

**Comment by thynson** [comment]
+1

**Comment by thecaddy** [comment]
+1

**Comment by kuoder** [comment]
+1

**Comment by intellix** [comment]
Coming from PHP/Doctrine and am evaluating sequelize for use with GraphQL. Class Table Inheritance is one of the items I'm currently using and thought maybe it might help to give a use-case to promote it's importance. 
I'm making a web game in which there are lots of future events happening: Troop Movements, Buildings, Units Training, Researches etc. I've got a cron/queue-worker which goes through these events and processes them in order of them completing.

```php
/**
 * @ORM\Table(name="game_events")
 * @ORM\InheritanceType("JOINED")
 * @ORM\DiscriminatorColumn(name="event", type="integer")
 * @ORM\DiscriminatorMap({
 *     1 = "Movement",
 *     2 = "BuildingQueue"
 * })
 * @ORM\Entity
 */
abstract class GameEvent {
    /** @ORM\GeneratedValue(strategy="AUTO") */
    private $id;
    /** @ORM\Column(name="start_time", type="datetime", nullable=false)  */
    private $startTime;
    /** @ORM\Column(name="end_time", type="datetime", nullable=true) */
    private $endTime;
}
```

```php
/**
 * @ORM\Table(name="building_queue")
 * @ORM\Entity
 */
class BuildingQueue extends GameEvent
{
    /** @ORM\OneToOne(targetEntity="Building", inversedBy="buildingQueue") */
    private $building;
}
```

```php
/**
 * @ORM\Table(name="movements")
 * @ORM\Entity
 */
class Movement extends GameEvent
{
    /** @ORM\ManyToOne(targetEntity="MapSquare") */
    private $mapSquare;
}
```

This creates a DB structure like:

```mysql
game_events
+----+---------------------+---------------------+-------+
| id | start_time          | end_time            | event |
+----+---------------------+---------------------+-------+
| 11 | 2016-03-22 21:24:38 | 2016-03-22 21:25:34 |     1 |
| 12 | 2016-03-22 22:35:06 | 2016-03-22 22:23:32 |     2 |
+----+---------------------+---------------------+-------+

movements
+----+---------------+
| id | map_square_id |
+----+---------------+
| 11 |         68874 |
+----+---------------+

building_queue
+----+---------------+
| id |   building_id |
+----+---------------+
| 12 |            22 |
+----+---------------+
```

So this means my queue-worker can process game-events in correct order:

```php
$gameEvents = $gameEventRepository->findAll(); // [ Movement, BuildingQueue ] 
```

And within the ORM they're regarded as separate entities: `$building->getBuildingQueue()` 

Is this any closer to inception? hasn't been an update for a year. Maybe this could be achieved with TypeScript classes and decorators?

**Comment by mickhansen** [comment]
@intellix No one has worked on this as far as i know.

**Comment by thomas-p-wilson** [comment]
I've been working on a quick and dirty plugin to mimic table-per-class inheritance, in order to get me where I'm going. On my journey, I noticed that each model attribute specifies the model from which it originated, which as far as I can tell never differs from the model that declared the attribute in the first place.

For example, if I inspect `user.attributes` where `user` is a model, I may receive the following result:

```
{
    first_name: {
        type: {
            options: [Object],
            _binary: undefined,
            _length: 255
        },
        allowNull: false,
        Model: user,
        fieldName: 'first_name',
        _modelAttribute: true,
        field: 'first_name'
    }
}
```

Note the `Model: user` attribute. The way I see it, if we declare that a model inherits from another model, we could merely append the parent model fields to the child model, as they would ideally retain that `Model: <model>` attribute. As a result, through the entire process of building the query, we would know which models were involved.

So now that we understand which models are used in generating our query, we need only distinguish the types of inheritance:

**Single Table Inheritance**: The quickest win I can think of in order to get this working is to apply the parent attributes to the child, and then set the child's table to that of the parent. This results in a select statement devoid of joins or other trickery, that pulls data from the table specified on the parent, and dumps that data into the child directly.

**Table Per Concrete Class**: When building the query, we need to peek at that `Model: <model>` attribute on each field. If there are attributes from two or more tables, we need to check our inheritance type attribute/indicator and if it's set to "table per class hierarchy", then we need to group the fields by owning model, join on those models, and pull the data in accordingly. Again, all the data gets dumped into the child model.

**Arbitrary Schema Design**: For those cases where we end up storing data in separate tables where the tables do not necessarily correspond 1:1 with a model, we should provide the ability to specify which table an attribute comes from. The result would be roughly the same as the implementation for "Table Per Concrete Class", but would allow the developer to alter the source table for an attribute.

In terms of configuration, I have two proposals:

1) Handle this with configuration properties. For example:
```
sequelize.define('client', { ...attributes... }, {
    inheritsFrom: sequelize.model('user'),
    inheritanceType: 'table-per-class'
}
```

2) Handle this in the same way we handle relations:
```
var Client = sequelize.define('client', { ...attributes... });
Client.inheritsFrom(sequelize.model('User'), {
    inheritanceType: 'single-table'
})
```

I expect that 'single table' and 'table per concrete class' inheritance types probably meet the needs of most folks looking to use this library. I can definitely see this working well, but obviously at this point I'm just an interloper. If this rough concept seems reasonable, I'm game for attempting implementation. I don't have a great deal of time at the moment, but it's something I need myself anyway. Thanks for your consideration.

**Comment by felixfbecker** [comment]
As we are going for classes in Sequelize 4 imo two models that inherit from each other should actually inherit from each other, like
```js
class Base extends Sequelize.Model {}
Base.init(...)
class Derived extends Base {}
Derived.init(...)
```
Upon the `init()` call this could be detected by walking the prototype chain until we reach `Sequelize.Model`.

What's important here is that we ensure a deep inheritance: Normally the model would inherit all properties, but the properties themselves not (`attributes`/`options` would just get overwritten). So what we have to do is something like 
```js
init() {
  this.attributes = Object.assign(Object.create(this.attributes), attributes)
}
```
which would make `attributes` inherit all properties from it's parent model aswell.

**Comment by thomas-p-wilson** [comment]
Guess I should have figured that this wouldn't be backported. I'm on 3.x right now, so I was mostly speaking based on what I know of that. Haven't the foggiest clue about 4.x. Either way, I'm afraid that if I can't get inheritance working to some degree for my current project, I'll have to move on, which is disappointing. This thread has been around for two and a half years.

**Comment by mickhansen** [comment]
@thomas-p-wilson Do you need to inherit attributes or everything including associations and methods?

**Comment by felixfbecker** [comment]
> I'll have to move on, which is disappointing.

It is disappointing that you may have to update to get new features?

> This thread has been around for two and a half years.

Because noone found it important enough to take the time to implement it 😐

**Comment by thomas-p-wilson** [comment]
Go easy there. Never said I didn't want to update. I'm fine with updating, but I'll not be updating to a pre-release in a production environment. At any rate, unless I'm mistaken, the inheritance functionality doesn't exist in 4.x at the moment.

And yes, nobody found it important enough to implement, I get that. I'm just surprised that nobody has bothered with any sort of inheritance in the two years this thread has been around. Makes me wonder if folks have been jumping ship for some other solution that offers it. I'm willing to give of my time in order to attempt an implementation, but I can't update to 4.x in the application I'm working on, ergo it'll be spare time work for me.

**Comment by thomas-p-wilson** [comment]
@mickhansen Attribute inheritance would be sufficient. I can get away without association and method inheritance

**Comment by felixfbecker** [comment]
@thomas-p-wilson yes, nothing is implemented so far in v4, besides that models are actual classes and can inherit on a very basic javascript level. Is your concern supporting actual table inheritance with `INHERITS FROM` or do you just want inheritance for your models?

**Comment by thomas-p-wilson** [comment]
I'm not overly concerned with `INHERITS FROM` support at present. Obviously it'd be nice, but it makes the most sense to first achieve inheritance support across all our supported datastores, no?

**Comment by felixfbecker** [comment]
That would be pretty easy if we get the Javascript-level inheritance support working. Wherever the library references `attributes` etc. It would just prototypically inherit from the parent's attribute.

**Comment by thomas-p-wilson** [comment]
Yep. Thinking I'll pull 4.x down tonight or tomorrow evening and muck about with it a bit.

**Comment by mickhansen** [comment]
I believe there's an issue with cloning attributes that probably makes it unworkable currently, but otherwise attribute sharing would be easy in userland.

**Comment by thomas-p-wilson** [comment]
Yeah I mucked about with attribute cloning in 3.x, but didn't go too far with it after realizing I'd have to change a whole bunch of stuff.

I'll get familiar with 4.x over the next couple days and noodle on the matter some more.

**Comment by felixfbecker** [comment]
@mickhansen what attribute cloning?
@thomas-p-wilson v4 really hasn't changed much besides that everything is ES6 and Model and Instance are one class.

**Comment by mickhansen** [comment]
@felixfbecker If you attempt to reuse an attribute definition and use a deep clone or similar it will fail.

```js
const mySharedAttributes = {
  firstName: Sequelize.STRING
};

sequelize.define('User', {
  ..._.deepClone(mySharedAttributes)
});

sequelize.define('Guest', {
  ..._.deepClone(mySharedAttributes)
});
```

IIRC this fails because type definitions end up behaving incorrectly.

**Comment by felixfbecker** [comment]
@mickhansen This is because `Sequelize.STRING` is a function, and you cannot clone it (or it wouldn't make sense to do so). But what do you think about my idea of adding to code to `init()` that makes `attributes` and `options` prototypically inherit from the parent model?

**Comment by ewyler** [comment]
I've been fighting for an elegant solution to this in Sequelize and, after many hours, haven't come up with a good solution aside from manually mapping tons of relations and attributes. I'm not sure why, but it unfortunately seems that the only library which supports generalization hierarchies is node-orm2 (https://github.com/dresende/node-orm2/wiki/extendsTo); unfortunately, it's not very widely adopted yet so that's a risky dependency to take on. I'll keep looking, but it'd be really awesome if Sequelize supported this - other ORMs, like SQLAlchemy, provide some working version of it (linked earlier in the thread).

It would be great to support the correct classifications if it is implemented as well - specifically, disjoint and overlapping hierarchies. I haven't found a library that does this, but it's common in RDBMS literature

**Comment by ewyler** [comment]
I take part of my comment back about node-orm2 - I misread it, that's actually still a `hasOne` association under the hood, not an inheritance hierarchy.

**Comment by thecaddy** [comment]
@ewyler I wrote a plugin for sequelize a year ago that is pretty straight forward, we've been using it since.  It maintains all syntax, the only catch is the constructors are slightly different from traditional sequelize.

https://gist.github.com/thecaddy/b72a062b62b835fa06a8

https://github.com/thecaddy/sequelize-inheritance

**Comment by ewyler** [comment]
Yeah, INHERITS FROM is an interesting postgres feature but unfortunately not a real generalization hierarchy - and you lose constraints in child tables.

**Comment by StyleT** [comment]
Must-have feature from my point of view

**Comment by jawadakram20** [comment]
@mickhansen Any progress regarding single table inheritance??

**Comment by mickhansen** [comment]
@jawadakram20 I'm no longer an active maintainer on this project

**Comment by johncade** [comment]
Why was this closed? Seems like a valid request

**Comment by mickhansen** [comment]
@johncade It's not closed?

**Comment by jawadakram20** [comment]
@johncade  It is not closed and they have not worked on it as well. Its been years and discussion is longer but of no use I guess.

**Comment by abelosorio** [comment]
I've not read all this discussion but I think it's a must-have feature, at least when using PostgreSQL.

**Comment by sbonami** [comment]
I've coded a custom multi-table inheritance that may be appropriate for a library, but the API is ugly and I'd pretty well guarantee I haven't covered all the scenarios it should cover. Having a sanctioned & approved solution would be greatly greatly appreciated

**Comment by abelosorio** [comment]
This is still an issue. A big one!

**Comment by ahmadzahabi1** [comment]
I recommend to implement this feature in the future.

**Comment by schieck0** [comment]
Die sequelize. F***** supporters

**Comment by LarissaGuder** [comment]
Almost seven years.
Any predictions to be implemented, or is not in scope for now?
